### PR TITLE
Add hybrid collaborative filtering recommender

### PR DESCRIPTION
**Pull request by eu9ene**
# Goal

Integrate experimental hybrid collaborative filtering recommender. See https://github.com/Pocket/dl-metaflow-jobs/blob/main/home_personalization_jobs/HybridCFModelFlow.py

## Todos
- [x] Deploy on dev
- [ ] Deploy in prod under the feature flag to test manually
- [ ] Make sure the logic satisfies our expectations
- [ ] More tests if needed
- [x] Load test

## Reference

Tickets:
* [Link to JIRA tickets](https://getpocket.atlassian.net/browse/DIS-604)

## Implementation Decisions
- use our regular S3 bucket for model storage instead of Metaflow one. Metaflow lib is very heavy and brings too many dependencies
- implemented a custom slate provider to not mess with already overwhelmed abstraction in ForYou provider

### CODE REVIEW

**Review by n0coast — APPROVED** [review]

**Review by mmiermans — APPROVED** [review]
Only minor suggestions. Nice work integrating an experiment in Recommendation API! It's good to have more people familiar with this code base. :muscle:

**Comment by eu9ene** [comment]
I ran a load test on dev with 20RPS and I don't see a big difference with the default home. I do see a higher CPU utilization (~60%) and ECS starts scaling the service. Our real load in production is way lower (~20RPM), so I don't anticipate issues here, at least until we land all users on home. I also increased memory to accommodate model reloading.

File: app/config.py [hunk 1]
```diff
@@ -99,3 +98,13 @@
     '058011b8-c70d-4a25-92e5-478e3ff0f0e6',  # Science
     '45f8e740-42e0-4f54-8363-21310a084f1f',  # Self-improvement
 ]
+
+hybrid_cf = {
+    'model_loader': os.getenv('HYBRID_CF_MODEL_LOADER', 's3'),  # or local,
+    'model_bucket': os.getenv('HYBRID_CF_MODEL_BUCKET', 'pocket-data-learning'),
+    'model_artifacts': os.getenv('HYBRID_CF_MODEL_ARTIFACTS',
+                                 'home_personalization/models/model_hybrid_cf/model_artifacts_hybrid_cf_recsys_with_tags_itemsWith5Users.pkl'),
```
Comment (chidamnat) [inline app/config.py:106]: good for now. But we will change the name of the model artifact and folder structure a bit. But now waiting on the SRE issue

File: app/recommenders/hybrid_cf.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        user_idx = model.users2idx.get(user_id)
+        if user_idx is None:
+            logging.info(f'HybridCF: user not found {user_id}')
+            return None
+
+        # run inference
+        dot_prod = np.dot(model.item_factors, model.user_factors[user_idx])
+        num_approved = len(model.approved_subset_items)
+        scores = (dot_prod + model.items_bias + model.users_bias[user_idx])[:num_approved]
```
Comment (chidamnat) [inline app/recommenders/hybrid_cf.py:84]:
```suggestion
        scores = (dot_prod + model.items_bias + model.users_bias[user_idx])
```

Comment (chidamnat) [reply app/recommenders/hybrid_cf.py:84]: please refer to this https://github.com/Pocket/data-explorations/blob/home_personalization/home_personalization/prediction_example_hybridCF.ipynb

**Review by chidamnat — APPROVED** [review]
lgtm. thanks Evgeny!

**Pull request merged**